File: firebase/functions/lib/admin/payments/get_stripe_connected_account_link.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import '../../utils/infra/firestore_utils.dart';
import 'stripe_util.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/admin/partner_agreement.dart';

class GetStripeConnectedAccountLink
    extends OnCallMethod<GetStripeConnectedAccountLinkRequest> {
  GetStripeConnectedAccountLink()
      : super(
          'getStripeConnectedAccountLink',
          (json) => GetStripeConnectedAccountLinkRequest.fromJson(json),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetStripeConnectedAccountLinkRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(context.authUid != null);

    final agreementRef =
        firestore.document('partner-agreements/${request.agreementId}');
    final agreementDoc = await agreementRef.get();
    orElseUnauthorized(agreementDoc.exists);

    final agreement = PartnerAgreement.fromJson(agreementDoc.data.toMap());
    orElseUnauthorized(agreement.allowPayments == true);
    orElseUnauthorized(agreement.stripeConnectedAccountId != null);
    orElseUnauthorized(agreement.initialUserId == context.authUid);

    final domain = functions.config.get('app.domain') as String;

    final Map<String, String> params = {
      'account': agreement.stripeConnectedAccountId!,
      'type': 'account_onboarding',
      'return_url': 'https://$domain/${request.responsePath}',
      'refresh_url': 'https://$domain/${request.responsePath}',
    };

    final jsonResponse =
        await stripeUtil.post(path: '/account_links', params: params);
    return GetStripeConnectedAccountLinkResponse(url: jsonResponse['url'])
        .toJson();
  }

  // Sets minimum instances to 0
  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions
        .runWith(
          runWithOptions ??
              RuntimeOptions(
                timeoutSeconds: 60,
                memory: '1GB',
                minInstances: 0,
              ),
        )
        .https
        .onCall(callAction);
  }
}

File: firebase/functions/src/admin/payments/get_stripe_connected_account_link.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { stripeUtil } from './stripe_util';
import { orElseUnauthorized } from '../../utils/utils';

interface GetStripeConnectedAccountLinkRequest {
  agreementId: string;
  responsePath: string;
}

export class GetStripeConnectedAccountLink extends OnCallMethod<GetStripeConnectedAccountLinkRequest> {
  constructor() {
    super(
      'getStripeConnectedAccountLink',
      (json) => json as GetStripeConnectedAccountLinkRequest,
      { runWithOptions: { timeoutSeconds: 60, memory: '1GB', minInstances: 0 } }
    );
  }

  async action(request: GetStripeConnectedAccountLinkRequest, context: functions.https.CallableContext): Promise<Record<string, unknown>> {
    orElseUnauthorized(context.auth?.uid != null);

    const agreementRef = firestore.doc(`partner-agreements/${request.agreementId}`);
    const agreementDoc = await agreementRef.get();
    orElseUnauthorized(agreementDoc.exists);

    const agreement = firestoreUtils.fromFirestoreJson(agreementDoc.data() ?? {}) as Record<string, unknown>;
    orElseUnauthorized(agreement['allowPayments'] === true);
    orElseUnauthorized(agreement['stripeConnectedAccountId'] != null);
    orElseUnauthorized(agreement['initialUserId'] === context.auth?.uid);

    const domain = functions.config().app?.domain as string;

    const params: Record<string, string> = {
      'account': agreement['stripeConnectedAccountId'] as string,
      'type': 'account_onboarding',
      'return_url': `https://${domain}/${request.responsePath}`,
      'refresh_url': `https://${domain}/${request.responsePath}`,
    };

    const jsonResponse = await stripeUtil.post({ path: '/account_links', params });
    return { url: jsonResponse['url'] };
  }
}
